From 8b4f5ec41c0f073627953fe5f80d3b9f3527f9b5 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:10:05 -0400
Subject: [usb] ehci: avoid BIOS handover on the HASEE E200

Message-id: <1410887463-20674-534-git-send-email-dzickus@redhat.com>
Patchwork-id: 95392
O-Subject: [RHEL7 PATCH 533/591] USB: EHCI: avoid BIOS handover on the HASEE E200
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit b0a50e92bda3c4aeb8017d4e6c6e92146ebd5c9b
Author: Alan Stern <stern@rowland.harvard.edu>
Date:   Tue Jun 3 11:00:27 2014 -0400

    USB: EHCI: avoid BIOS handover on the HASEE E200

    Leandro Liptak reports that his HASEE E200 computer hangs when we ask
    the BIOS to hand over control of the EHCI host controller.  This
    definitely sounds like a bug in the BIOS, but at the moment there is
    no way to fix it.

    This patch works around the problem by avoiding the handoff whenever
    the motherboard and BIOS version match those of Leandro's computer.

    Signed-off-by: Alan Stern <stern@rowland.harvard.edu>
    Reported-by: Leandro Liptak <leandroliptak@gmail.com>
    Tested-by: Leandro Liptak <leandroliptak@gmail.com>
    CC: <stable@vger.kernel.org>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/host/pci-quirks.c b/drivers/usb/host/pci-quirks.c
index 4a6d3dd..2f3aceb 100644
--- a/drivers/usb/host/pci-quirks.c
+++ b/drivers/usb/host/pci-quirks.c
@@ -656,6 +656,14 @@ static const struct dmi_system_id ehci_dmi_nohandoff_table[] = {
    DMI_MATCH(DMI_BIOS_VERSION, "Lucid-"),
   },
  },
+ {
+  /* HASEE E200 */
+  .matches = {
+   DMI_MATCH(DMI_BOARD_VENDOR, "HASEE"),
+   DMI_MATCH(DMI_BOARD_NAME, "E210"),
+   DMI_MATCH(DMI_BIOS_VERSION, "6.00"),
+  },
+ },
  { }
 };
 
@@ -665,9 +673,14 @@ static void ehci_bios_handoff(struct pci_dev *pdev,
 {
  int try_handoff = 1, tried_handoff = 0;
 
- /* The Pegatron Lucid tablet sporadically waits for 98 seconds trying
-  * the handoff on its unused controller.  Skip it. */
- if (pdev->vendor == 0x8086 && pdev->device == 0x283a) {
+ /*
+  * The Pegatron Lucid tablet sporadically waits for 98 seconds trying
+  * the handoff on its unused controller.  Skip it.
+  *
+  * The HASEE E200 hangs when the semaphore is set (bugzilla #77021).
+  */
+ if (pdev->vendor == 0x8086 && (pdev->device == 0x283a ||
+   pdev->device == 0x27cc)) {
   if (dmi_check_system(ehci_dmi_nohandoff_table))
    try_handoff = 0;
  }
-- 
1.7.1